### PR TITLE
Extra methods to facilitate icons refresh

### PR DESCRIPTION
**Pull request by ghybs**
Additional methods to facilitate marker icons update and propagate to their parent clusters in real-time. For issues #563, #561, #555, #535 and #498.

The 2 private methods could also be re-used in `addLayers` and `removeLayers`.

As I am quite new to Git, I am not sure this is the correct way to proceed. I also tried to find a suitable place for this piece of code, but feel free to place it somewhere else more appropriate.

Anyway, this is mainly an attempt to propose some code in a better way than posting it in comments, as pointed out by @IvanSanchez.

### CODE REVIEW

**Comment by danzel** [comment]
I like the idea of this.
Could you please have a go at adding some tests to the spec dir to cover this behaviour?
You should also update the build script to include this by default.

Thanks!

**Comment by ghybs** [comment]
Hi,

Thanks for your feedback.
Sure, will do as soon as I have time.
I will also update the readme accordingly.

**Comment by ghybs** [comment]
Hi,

Excuse me I have some doubts about how I should proceed here:
- Should I update the dist files as well, so that everyone can benefit from the new API right away? Or should I leave those unchanged, and let you do that once you consider the src files stable enough? In particular, updating the dist files would include all previous merges that are not part of the current dist files.
- Should I bump up the minor version in bower and package.json files? Because there **is** a new API, that is backward compatible (it does not break anything, at least for the methods I introduce).
- Should I update the README file to communicate about the new methods?

I apologize for these stupid questions, I am quite new to GitHub and I am not so aware of how an open source community runs…

**Comment by danzel** [comment]
No worries, most projects run differently :)

Don't update the dist files (so please revert that commit)
Don't update the version number (that is done when we do a release).

Do update the README to detail this new method :)

**Comment by ghybs** [comment]
Hi,

Thank you for the explanations.
I will proceed as instructed.

By the way, please how do you manage the discrepancy between Readme and
version of dist file?
If I update the Readme, it will talk about something that is not in the
dist version.

**Comment by danzel** [comment]
I still see the dist changes in your commits.

I'll update dist after merging.

**Comment by ghybs** [comment]
Hi,

Sorry for the delay.

I am not sure how to revert a commit…
I will try rebuilding the fork.

**Comment by ghybs** [comment]
Ok it should be good now.
I actually just copied back the original dist files, I hope this trick is
enough?

Please let me know should I need to perform any extra step or correct any
mistake.

**Comment by danzel** [comment]
That looks good, will take a read through next week.

File: src/MarkerClusterGroup.Refresh.js [hunk 1]
```diff
@@ -0,0 +1,98 @@
+/**
+ * Adds 1 public method to MCG and 1 to L.Marker to facilitate changing
+ * markers' icon options and refreshing their icon and their parent clusters
+ * accordingly (case where their iconCreateFunction uses data of childMarkers
+ * to make up the cluster icon).
+ * Should cover issues #563, #561, #555, #535 and #498.
```
Comment (danzel) [inline src/MarkerClusterGroup.Refresh.js]: Trash this line please, we have git history and github to track this :)

**Comment by danzel** [comment]
Merged, can trash the line myself

**Comment by ghybs** [comment]
Hi,

Well it looks like the test spec messes up…
Quite often it hangs the test process, even though when it succeeds, everything passes.

**Comment by danzel** [comment]
Dumb, this happens in some of the current tests already too.

**Comment by ghybs** [comment]
It seems that avoiding `beforeEach` and `afterEach` and trying to re-use objects whenever possible makes the test process much more reliable (even though still not 100% perfect in my tests).

It could be related to PhantomJS bad garbage collection / memory leak, even though we are still on a very low number of tests and I do not see any peak in memory...

I will push an update for the test I created at least.

**Comment by ghybs** [comment]
See PR #577.
It fixes the issue on my machine at least.

Maybe re-engineering the other test suites could help for the similar cases you sound to have encountered?

**Comment by danzel** [comment]
sweet, next time I see any others crap out I'll have a go.

**Comment by ayozebarrera** [comment]
Can't find refreshClusters() method anywhere. What is the state of this at the moment? I need to update some markers every few seconds

**Comment by ghybs** [comment]
Hi,

It should be included in the dist file.
Which branch / Leaflet version are you using?

For Leaflet stable (0.7.7), use marker cluster [branch "leaflet-0.7"](https://github.com/Leaflet/Leaflet.markercluster/tree/leaflet-0.7), `refreshClusters` is [there](https://github.com/Leaflet/Leaflet.markercluster/blob/leaflet-0.7/dist/leaflet.markercluster-src.js#L2375).

For Leaflet 1.0 beta2, use marker cluster [tag "1.0.0-beta.2.0"](https://github.com/Leaflet/Leaflet.markercluster/tree/v1.0.0-beta.2.0), `refreshClusters` is [there](https://github.com/Leaflet/Leaflet.markercluster/blob/v1.0.0-beta.2.0/dist/leaflet.markercluster-src.js#L2430).

Hope this helps.

**Comment by ayozebarrera** [comment]
Nice, it helps.

I'm using leaflet 0.7.7 but I'm using default branch of marker cluster. How I switch branch?

Thanks for the quick reply :p

**Pull request merged**